spec.html

```diff
@@ -556,12 +560,8 @@ <h1><ins>GetCycleRoot ( _module_ )</ins></h1>
       <h1><ins>CyclicModuleExecutionFulfilled ( _module_ )</ins></h1>
       <emu-alg>
         1. Assert: _module_.[[EvaluationError]] is *undefined*.
-        1. Assert: _module_.[[Status]] is `"evaluating"`, `"evaluating-async"` or `"evaluated"`.
-        1. If _module_.[[Status]] is `"evaluating"`, then
-          1. Assert: _module_.[[Async]] is *false*.
-          1. Assert: _module_.[[AsyncParentModules]] is an empty List.
-        1. If _module_.[[Status]] is `"evaluating-async"`, then
-          1. Set _module_.[[Status]] to `"evaluated"`.
+        1. Assert: _module_.[[Status]] is `"evaluating-async"`.
```

Comment:
Reconsider my example with two mutually dependent async modules. We seem to still end up here with [[Status]] set to "evaluation" with the stack Evaluate() -> InnerModuleEvaluation() -> InnerModuleEvaluation() -> ExecuteCyclicModule() -> CyclicModuleExecutionFulfilled().

Swapping the steps 
```markdown
             1. <ins>Perform ! ExecuteCyclicModule(_module_).</ins>
             1. <ins>Set _module_.[[Status]] to `"evaluating-async"`.</ins>
```

will get us past this assertion, but then we set [[Status]] to "evaluated" without removing it from the stack, and that messes up the other assertion.

Reply:
CyclicModuleExecutionFulfilled will always be a promise resolution in an async cycle, and as a result will always run asynchronously and thus after we have set the state to `"evaluating-async"`.

In the case where the bottom of the cycle is synchronous, we instead use the evaluate form that throws its completion to InnerModuleEvaluation, such that the stack reporting works out for those sync cases.

So I believe we are fine - but please clarify further if you can.

Reply:
Looks like I got this wrong indeed.

However, there's still an issue for this case. Let's say A <-> B and we call A.Evaluate(). We then end up with B.[[Status]] set to evaluating-async, and calling GetCycleRoot(B) in the steps below, while B.[[AsyncParentModules]] is empty, which asserts.

```markdown
            1. If _requiredModule_.[[Status]] is `"evaluating"`, then
              1. Set _module_.[[DFSAncestorIndex]] to min(_module_.[[DFSAncestorIndex]], _requiredModule_.[[DFSAncestorIndex]]).
            1. <ins>Otherwise, set _requiredModule_ to GetCycleRoot(_requiredModule_).</ins>
```

Reply:
Completely correct. Closing to rather split out the fixes and include a fix along the lines provided in https://github.com/tc39/proposal-top-level-await/issues/89#issuecomment-497766800.